### PR TITLE
Fixed URL pre-selection of configurable product swatches with associated product images throws JavaScript error #18017 In Magento 2.2

### PR DESCRIPTION
**Pull request by niravkrish**
### Description (*)

When pre-selecting a configurable product swatch that has an image, Magento will try to load that image in the product gallery before it is initialized, throwing a JavaScript error.

### Fixed Issues (if relevant)

1. magento/magento2#18017: URL pre-selection of configurable product swatches with associated product images throws JavaScript error

### Manual testing scenarios (*)

1. Hit Url With Url domain.com/montana-wind-jacket.html#93=53
2. Test with normal configurable product on listing page as well as on Product Details Page

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @niravkrish. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.2-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

File: app/code/Magento/Swatches/view/frontend/web/js/swatch-renderer.js [hunk 1]
```diff
@@ -1166,7 +1166,7 @@ define([
         updateBaseImage: function (images, context, isInProductView, eventName) {
             var gallery = context.find(this.options.mediaGallerySelector).data('gallery');
 
-            if (eventName === undefined) {
+            if (eventName === undefined && gallery !== undefined) {
```
Comment (VladimirZaets) [inline app/code/Magento/Swatches/view/frontend/web/js/swatch-renderer.js]: Hi @niravkrish, thanks for collaboration. Please use `typeof` (typeof gallery === undefined) or cast value to bolean uses !! (!!galery).

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-3751](https://jira.corp.magento.com/browse/ENGCOM-3751) has been created to process this Pull Request

**Review by DanielRuf — CHANGES_REQUESTED** [review]
Please use `"undefined"`.

**Comment by niravkrish** [comment]
Hello @DanielRuf 

undefined is a keyword. as per your suggestion "undefined" is treat as string.

so please give me reason to change undefined to "undefined"

**Comment by DanielRuf** [comment]
> undefined is a keyword

This is not correct. There is no variable `undefined`, please see the strict cide on JavaScript. Also `typeof` returns a string, eg `"undefined"`.

**Comment by DanielRuf** [comment]
> undefined is a keyword

This is not correct. There is no variable `undefined`, please see the strict mode in JavaScript. Also `typeof` returns a string, eg `"undefined"`.

**Comment by sdzhepa** [comment]
Hello @niravkrish 
During testing our QA engineer faced an issue.
Could you please have a look and fix it?

Please see the quote of QA comment after testing on branch with the PR

**Problem:** _PR doesn't fix issue #18017_

**Manual testing scenario:**

1. Create two option for configurable product based on Visual Swatch attribute
2. Set Use Web Server Rewrites=No
3. Open the configurable product page, with the option pre-selected.
` magento_site/index.php/conf1.html#135=4`

**Actual Result:** Not OK,
19635pr




Thank you for contribution and collaboration

**Comment by niravkrish** [comment]
I have already change code suggested by reviewer.
 
Please review it and let me know for further changes.

**Comment by sivaschenko** [comment]
Hi @niravkrish can you please merge the mainline and resolve conflicts.
Looks like there is a partial resolution of the issue already present in mainline.

**Comment by sivaschenko** [comment]
Hi @niravkrish thanks for branch update. I wonder if issue was fixed on fresh 2.2-develop? It contained very similar code to this pull request that was replaced by the merge commit: https://github.com/magento/magento2/pull/19627/commits/c25be183d3af9ccc3e9d1fb4688bb88df2a6829b

**Comment by niravkrish** [comment]
I think someone already solve this issue.
so I am closing this ticket.

**Pull request closed without merging**